### PR TITLE
[SYCLomatic]Refine function implementation of checkTemplateArgSpelling

### PR DESCRIPTION
**Pull request by wenhuiNi**
1. Fix irregular variable naming.
2. Reduce the number of nesting layers for if-else

Signed-off-by: Ni, Wenhui <wenhui.ni@intel.com>

### CODE REVIEW

File: clang/lib/DPCT/TypeLocRewriters.cpp [hunk 1]
```diff
@@ -91,35 +91,37 @@ class CheckTypeNameAndInVarDecl {
...
-            const ValueDecl *decl = arg.getAsDecl();
-            return decl->getNameAsString() == str;
-          } else if (arg.getKind() == TemplateArgument::ArgKind::Integral) {
-            const llvm::APSInt &value = arg.getAsIntegral();
-            return std::to_string(value.getExtValue()) == str;
-          } else if (arg.getKind() == TemplateArgument::ArgKind::Expression) {
-            const Expr *expr = arg.getAsExpr();
-            return getStmtSpelling(expr) == str;
+          const TemplateArgument &TARG = TST->template_arguments()[Index];
+          if (TARG.getKind() == TemplateArgument::ArgKind::Type) {
```
Comment (yihwang-nv) [inline clang/lib/DPCT/TypeLocRewriters.cpp]: What about to use a `switch` here?

File: clang/lib/DPCT/TypeLocRewriters.cpp [hunk 2]
```diff
@@ -91,35 +91,37 @@ class CheckTypeNameAndInVarDecl {
 };
 
 inline std::function<bool(const TypeLoc TL)>
```
Comment (yihwang-nv) [inline clang/lib/DPCT/TypeLocRewriters.cpp:93]: The return type can be `auto`.

**Review by tomflinda — APPROVED** [review]
LGTM

File: clang/lib/DPCT/TypeLocRewriters.cpp [hunk 3]
```diff
@@ -91,37 +91,29 @@ class CheckTypeNameAndInVarDecl {
...
-checkTemplateArgSpelling(size_t index, std::string str) {
+checkTemplateArgSpelling(size_t Index, std::string Str) {
+  auto getQualtifiedNameStr = [=](const NamedDecl *NL) -> std::string {
+    if (const auto *NSD = dyn_cast<NamespaceDecl>(NL->getDeclContext())) {
```
Comment (ziranzha) [inline clang/lib/DPCT/TypeLocRewriters.cpp:96]: Need check NL is NULL at first.

File: clang/lib/DPCT/TypeLocRewriters.cpp [hunk 4]
```diff
@@ -91,37 +91,29 @@ class CheckTypeNameAndInVarDecl {
...
-            return getStmtSpelling(expr) == str;
-          }
-        }
+      if (Index > TSTL.getNumArgs())
+        return false;
+      const auto TA = TSTL.getArgLoc(Index).getArgument();
+      if (TA.getKind() == TemplateArgument::ArgKind::Type) {
+        return getQualtifiedNameStr(TA.getAsType()->getAsTagDecl()) == Str;
+      } else if (TA.getKind() == TemplateArgument::ArgKind::Declaration) {
+        return TA.getAsDecl()->getNameAsString() == Str;
```
Comment (ziranzha) [inline clang/lib/DPCT/TypeLocRewriters.cpp]: Does this also need the qualifier?

File: clang/lib/DPCT/TypeLocRewriters.cpp [hunk 5]
```diff
@@ -91,37 +91,29 @@ class CheckTypeNameAndInVarDecl {
...
 inline std::function<bool(const TypeLoc TL)>
-checkTemplateArgSpelling(size_t index, std::string str) {
+checkTemplateArgSpelling(size_t Index, std::string Str) {
+  auto getQualtifiedNameStr = [=](const NamedDecl *NL) -> std::string {
+    if (const auto *NSD = dyn_cast<NamespaceDecl>(NL->getDeclContext())) {
+      std::string TypeQualifiedString =
+          getNameSpace(NSD) + "::" + NL->getNameAsString();
+      return TypeQualifiedString;
+    }
+    return "";
```
Comment (ziranzha) [inline clang/lib/DPCT/TypeLocRewriters.cpp]: If there is no qualifier, the type name will always be empty, is it expected?

Comment (wenhuiNi) [reply clang/lib/DPCT/TypeLocRewriters.cpp]: It is mistake. Fixed

File: clang/lib/DPCT/TypeLocRewriters.cpp (hunk 4 above)
Comment (wenhuiNi) [reply clang/lib/DPCT/TypeLocRewriters.cpp]: Fixed

File: clang/lib/DPCT/TypeLocRewriters.cpp [hunk 6]
```diff
@@ -91,37 +91,29 @@ class CheckTypeNameAndInVarDecl {
 };
 
 inline std::function<bool(const TypeLoc TL)>
-checkTemplateArgSpelling(size_t index, std::string str) {
+checkTemplateArgSpelling(size_t Index, std::string Str) {
+  auto getQualtifiedNameStr = [=](const NamedDecl *NL) -> std::string {
```
Comment (ziranzha) [inline clang/lib/DPCT/TypeLocRewriters.cpp:95]: Need check `NL` is null at first?

**Review by ziranzha — APPROVED** [review]

**Pull request merged**